File: naked-generator/Primary/c/acu.c
#include <string.h>
#include <assert.h>
#include <stdio.h>

#include "acu.h"

void serialize_TS_STATUS(TS_STATUS* ts_status, uint8_t* buffer, size_t buf_len) {
	assert(buf_len >= sizeof(TS_STATUS));
	memcpy(buffer, ts_status, sizeof(STEER_STATUS));
}
void serialize_STEER_STATUS(STEER_STATUS* steer_status, uint8_t* buffer, size_t buf_len) {
	assert(buf_len >= sizeof(STEER_STATUS));
	memcpy(buffer, steer_status, sizeof(STEER_STATUS));
}
void serialize_SET_CAR_STATUS(SET_CAR_STATUS* set_car_status, uint8_t* buffer, size_t buf_len) {
	assert(buf_len >= sizeof(SET_CAR_STATUS));
	memcpy(buffer, set_car_status, sizeof(STEER_STATUS));
}

void deserialize_TS_STATUS(uint8_t* buffer, size_t buf_len, TS_STATUS* ts_status) {
	assert(buf_len >= sizeof(TS_STATUS));
	memcpy(ts_status, buffer, sizeof(STEER_STATUS));
}
void deserialize_STEER_STATUS(uint8_t* buffer, size_t buf_len, STEER_STATUS* steer_status) {
	assert(buf_len >= sizeof(STEER_STATUS));
	memcpy(steer_status, buffer, sizeof(STEER_STATUS));
}
void deserialize_SET_CAR_STATUS(uint8_t* buffer, size_t buf_len, SET_CAR_STATUS* set_car_status) {
	assert(buf_len >= sizeof(SET_CAR_STATUS));
	memcpy(set_car_status, buffer, sizeof(STEER_STATUS));
}

File: naked-generator/Primary/py/acu.py
from enum import Enum
from struct import pack, unpack
from collections import namedtuple


class Ts_Status(Enum):
	OFF = 0
	PRECHARGE = 1
	ON = 2
	FATAL = 3


class Traction_Control(Enum):
	OFF = 0
	SLIP_CONTROL = 1
	TORQUE_VECTORING = 2
	COMPLETE = 3


class Map(Enum):
	R = 0
	D20 = 1
	D40 = 2
	D60 = 3
	D80 = 4
	D100 = 5


class Car_Status_Set(Enum):
	IDLE = 0
	RUN = 1


TS_STATUS = namedtuple('TS_STATUS', 'ts_status')
TS_STATUS_schema = '<c'

STEER_STATUS = namedtuple('STEER_STATUS', 'traction_control map radio_on')
STEER_STATUS_schema = '<cc?'

SET_CAR_STATUS = namedtuple('SET_CAR_STATUS', 'car_status_set')
SET_CAR_STATUS_schema = '<c'


def serialize_TS_STATUS(struct: TS_STATUS) -> bytearray:
	return pack(TS_STATUS, *tuple(struct))


def serialize_STEER_STATUS(struct: STEER_STATUS) -> bytearray:
	return pack(STEER_STATUS, *tuple(struct))


def serialize_SET_CAR_STATUS(struct: SET_CAR_STATUS) -> bytearray:
	return pack(SET_CAR_STATUS, *tuple(struct))


def deserialize_TS_STATUS(buffer: bytearray) -> TS_STATUS:
	return TS_STATUS._make(unpack(TS_STATUS_schema, buffer))


def deserialize_STEER_STATUS(buffer: bytearray) -> STEER_STATUS:
	return STEER_STATUS._make(unpack(STEER_STATUS_schema, buffer))


def deserialize_SET_CAR_STATUS(buffer: bytearray) -> SET_CAR_STATUS:
	return SET_CAR_STATUS._make(unpack(SET_CAR_STATUS_schema, buffer))


